File: pluk/src/store/savedCommands.ts
import { Database } from "bun:sqlite";
import { homedir } from "os";
import { mkdirSync } from "fs";

const DATA_DIR = `${homedir()}/.pluk`;
mkdirSync(DATA_DIR, { recursive: true });

const db = new Database(`${DATA_DIR}/pluk.db`);

// Pre-selected shell commands for SSH integrations. The user curates these
// (via the REST API / UI); agents run them by name through `run_saved_command`.
// There is no allowlist — saved commands run unrestricted as the connecting SSH
// user, exactly like an ad-hoc command. The MCP confirm prompt is the only gate.
db.run(`
  CREATE TABLE IF NOT EXISTS saved_commands (
    id TEXT PRIMARY KEY,
    connection_id TEXT NOT NULL,
    name TEXT NOT NULL,
    command TEXT NOT NULL,
    working_dir TEXT,
    created_at TEXT NOT NULL DEFAULT (datetime('now')),
    UNIQUE(connection_id, name)
  )
`);

export interface SavedCommand {
  id: string;
  connection_id: string;
  name: string;
  command: string;
  working_dir: string | null;
  created_at: string;
}

export type SavedCommandInput = Omit<SavedCommand, "id" | "created_at">;

export function listSavedCommands(connectionId: string): SavedCommand[] {
  return db.query(`
    SELECT id, connection_id, name, command, working_dir, created_at
    FROM saved_commands
    WHERE connection_id = ?
    ORDER BY name
  `).all(connectionId) as SavedCommand[];
}

export function getSavedCommand(connectionId: string, name: string): SavedCommand | null {
  return db.query(`
    SELECT id, connection_id, name, command, working_dir, created_at
    FROM saved_commands
    WHERE connection_id = ? AND name = ?
  `).get(connectionId, name) as SavedCommand | null;
}

export function createSavedCommand(input: SavedCommandInput): SavedCommand {
  const id = crypto.randomUUID();
  db.query(`
    INSERT INTO saved_commands (id, connection_id, name, command, working_dir)
    VALUES (?, ?, ?, ?, ?)
  `).run(id, input.connection_id, input.name, input.command, input.working_dir);
  return { ...input, id, created_at: new Date().toISOString() };
}

export function deleteSavedCommand(connectionId: string, name: string): boolean {
  const result = db.query("DELETE FROM saved_commands WHERE connection_id = ? AND name = ?").run(connectionId, name);
  return result.changes > 0;
}

File: crates/pluk-store/src/saved_commands.rs
//! Saved commands: user-curated shell commands for SSH integrations, run by
//! name through the `run_saved_command` tool. There is no allowlist beyond
//! this table — saved commands run unrestricted as the connecting SSH user,
//! exactly like an ad-hoc command; the MCP confirm prompt is the only gate.
//!
//! Ids are full-dashed UUIDs (`crypto.randomUUID()` in the TypeScript
//! sources) — unlike integrations and groups, whose 16-char ids come from
//! [`crate::ids`].

use rusqlite::{OptionalExtension, Row, params};

use crate::Store;
use crate::error::Result;
use crate::models::SavedCommand;

/// Everything a caller supplies; `id` and `created_at` are minted here.
#[derive(Debug, Clone)]
pub struct SavedCommandInput {
    pub connection_id: String,
    pub name: String,
    pub command: String,
    pub working_dir: Option<String>,
}

fn map_saved_command(row: &Row<'_>) -> rusqlite::Result<SavedCommand> {
    Ok(SavedCommand {
        id: row.get(0)?,
        connection_id: row.get(1)?,
        name: row.get(2)?,
        command: row.get(3)?,
        working_dir: row.get(4)?,
        created_at: row.get(5)?,
    })
}

impl Store {
    pub fn list_saved_commands(&self, connection_id: &str) -> Result<Vec<SavedCommand>> {
        let conn = self.conn.lock().expect("store lock");
        let mut stmt = conn.prepare(
            "SELECT id, connection_id, name, command, working_dir, created_at FROM saved_commands
             WHERE connection_id = ? ORDER BY name",
        )?;
        let rows = stmt.query_map([connection_id], map_saved_command)?;
        Ok(rows.collect::<std::result::Result<Vec<_>, rusqlite::Error>>()?)
    }

    pub fn get_saved_command(
        &self,
        connection_id: &str,
        name: &str,
    ) -> Result<Option<SavedCommand>> {
        let conn = self.conn.lock().expect("store lock");
        Ok(conn
            .query_row(
                "SELECT id, connection_id, name, command, working_dir, created_at FROM saved_commands
                 WHERE connection_id = ? AND name = ?",
                [connection_id, name],
                map_saved_command,
            )
            .optional()?)
    }

    /// Save a command. A duplicate `(connection_id, name)` surfaces the UNIQUE
    /// violation to the caller.
    pub fn create_saved_command(&self, input: &SavedCommandInput) -> Result<SavedCommand> {
        let id = uuid::Uuid::new_v4().to_string();
        let conn = self.conn.lock().expect("store lock");
        conn.execute(
            "INSERT INTO saved_commands (id, connection_id, name, command, working_dir)
             VALUES (?, ?, ?, ?, ?)",
            params![
                id,
                input.connection_id,
                input.name,
                input.command,
                input.working_dir
            ],
        )?;
        Ok(conn.query_row(
            "SELECT id, connection_id, name, command, working_dir, created_at FROM saved_commands WHERE id = ?",
            [&id],
            map_saved_command,
        )?)
    }

    pub fn delete_saved_command(&self, connection_id: &str, name: &str) -> Result<bool> {
        let conn = self.conn.lock().expect("store lock");
        Ok(conn.execute(
            "DELETE FROM saved_commands WHERE connection_id = ? AND name = ?",
            [connection_id, name],
        )? > 0)
    }
}
